panel/panel.py

```diff
@@ -136,64 +128,50 @@ def run_update_panel_cmd(panel, text=None):
     panel.run_command(cmd, {'text': text, 'clear_sel': clear_sel})
 
 
-def format_row(lineno, error_type, dic):
-    lineno = int(lineno) + 1
-    start = dic['start'] + 1
-    msg = dic['msg'].rstrip()
-    tmpl = " {LINENO:>5}:{START:<4} {ERR_TYPE:7} {linter:>12}: {code:12} {MSG}"
-    return tmpl.format(
-        LINENO=lineno, START=start, ERR_TYPE=error_type, MSG=msg, **dic)
+def format_row(item):
+    line = item["line"] + 1
+    start = item["start"] + 1
+    tmpl = " {LINE:>5}:{START:<4} {error_type:7} {linter:>12}: {code:12} {msg}"
+    return tmpl.format(LINE=line, START=start, **item)
 
 
-def filter_ok(check_val, key, panel_filter):
+def passes_listing(error, panel_filter):
     """Check value against white- and blacklisting and return bool."""
     # check whitelisting
     if not panel_filter:
         return True
 
-    if key in panel_filter and check_val not in panel_filter[key]:
-        return False
-
-    # check blacklisting
-    exclude_key = "exclude_" + key
-    if exclude_key in panel_filter and check_val in panel_filter[exclude_key]:
-        return False
-
+    COMMAND_ARG_TO_ERROR_KEY = {
+        "types": "error_type",
+        "codes": "code",
+        "linter": "linter"
+    }
+
+    for key, check_val in COMMAND_ARG_TO_ERROR_KEY.items():
+        # check whitelisting
+        if key in panel_filter:
+            if error[check_val] not in panel_filter[key]:
+                return False
+
+        # check blacklisting
+        exclude_key = "exclude_" + key
+        if exclude_key in panel_filter:
+            if error[check_val] in panel_filter[exclude_key]:
+                return False
+
+    # if all checks passed return True
     return True
 
 
-def get_view_lines(view_dict, panel_filter):
-    view_lines = []
-    for lineno, line_dict in sorted(view_dict["line_dicts"].items()):
-        for error_type in WARN_ERR:
-            if not filter_ok(error_type, "types", panel_filter):
-                continue
-
-            err_dict = line_dict.get(error_type)
-            if not err_dict:
-                continue
-            items = sorted(err_dict, key=lambda k: k['start'])
-
-            items = [
-                item for item in items
-                if filter_ok(item['linter'], "linter", panel_filter) and
-                filter_ok(item['code'], "codes", panel_filter)
-            ]
-            for item in items:
-                view_lines.append((lineno, error_type, item))
-
-    return view_lines
-
-
 def fill_panel(window, update=False, **panel_filter):
-    errors = persist.errors.data
+
+    errors = persist.raw_errors
     if not errors:
         return
 
-    errors = filter_errors(window, errors)
-    errors = dedupe_views(errors)
+    errors_by_bid = get_window_raw_errors(window, errors, panel_filter)
 
-    path_dict, base_dir = create_path_dict(errors)
+    path_dict, base_dir = create_path_dict(window, errors_by_bid)
```

Comment:
create_path_dict expects `bids`as its second argument, so passing down a dict is confusing. Pass down `errors_by_bid.keys()`